src/district/server/smart_contracts.cljs

```diff
@@ -115,132 +107,134 @@
         (println (:name contract) contract-address (.toLocaleString gas-used)))
       contract)))
 
-
-(defn- deploy-smart-contract* [contract-key {:keys [:placeholder-replacements :arguments]
-                                             :as opts} callback]
-  (let [{:keys [:abi :bin] :as contract} (load-contract-files (contract contract-key) @smart-contracts)]
-    (when-not bin
-      (throw (js/Error. (str "Contract " contract-key " is missing bin"))))
-    (let [opts (merge {:data (str "0x" (link-contract-libraries @(:contracts @smart-contracts) bin placeholder-replacements))
-                       :gas 4000000}
-                      (when-not (:from opts)
-                        {:from (first (web3-eth/accounts @web3))})
-                      opts)
-          Contract (apply web3-eth/contract-new @web3 abi (into (vec arguments) [(select-keys opts [:from :to :gas-price :gas
-                                                                                                    :value :data :nonce
-                                                                                                    :condition])]))
-          tx-hash (aget Contract "transactionHash")
-          filter-id (atom nil)]
-
-      (if-not (fn? callback)
-        (handle-deployed-contract! contract-key contract abi tx-hash)
-        (reset!
-          filter-id
-          (web3-eth/filter
-            @web3
-            "latest"
-            (fn [err]
-              (when err
-                (callback err))
-              (try
-                (when-let [contract (handle-deployed-contract! contract-key contract abi tx-hash)]
-                  (web3-eth/stop-watching! @filter-id)
-                  ;; reset is needed, otherwise web3 crashes with "Can't connect to" on next request
-                  ;; Reasons unknown. Needed only because of deasync hack
-                  (web3/reset @web3)
-                  (callback nil contract))
-                (catch js/Error err
-                  (callback err))))))))))
-
-
-(def deploy-smart-contract-deasynced (deasync deploy-smart-contract*))
-
-(defn deploy-smart-contract! [& args]
-  (if (:auto-mining? @smart-contracts)
-    (apply deploy-smart-contract* args)
-    (apply deploy-smart-contract-deasynced args)))
-
+(defn deploy-smart-contract!
```

Comment:
I think we should add to the comment that this only works for ganache or executing from unlocked accounts.
I implemented a version that supports transactions signing but not sure if we want to add that.

Reply:
`opts` takes a {:from "0x..."} so it is kinda documented that there needs to be an unlocked account. I'll reiterate this in the README.